lightning/src/onion_message/messenger.rs

```diff
@@ -569,34 +645,61 @@ where
 			},
 		};
 
-		log_trace!(self.logger, "Sending onion message {}", log_suffix);
-
-		if let Err(e) = self.send_onion_message(path, contents, reply_path) {
-			log_trace!(self.logger, "Failed sending onion message {}: {:?}", log_suffix, e);
-			return;
+		match self.send_onion_message(path, contents, reply_path) {
+			Err(SendError::InvalidFirstHop(first_node_id, onion_message)) => {
+				let network_graph = self.network_graph.deref().read_only();
```

Comment:
Ok, let me go ahead and move `NetworkGraph` to `DefaultMessageRouter`. I guess the idea would be to add a `Vec<SocketAddress>` field to `OnionMessagePath`? I used `NodeAnnouncement` since it would give the user more information on deciding whether they want to connect to the node. But perhaps `SocketAddress`es are sufficient. And I suppose

Reply:
Ok, let me go ahead and move `NetworkGraph` to `DefaultMessageRouter`. I guess the idea would be to add a `Vec<SocketAddress>` field to `OnionMessagePath`? I used `NodeAnnouncement` since it would give the user more information on deciding whether they want to connect to the node. But perhaps `SocketAddress`es are sufficient. And I suppose a `MessageRouter` may want to connect to a private node, so we should support that.

> (and, awkwardly, passing some event through the peermanager to the user to trigger the connection, but really that is a router-originated event)

An alternative could be to parameterize `BackgroundProcessor` with `OnionMessenger` or `MessageRotuer`, but I figured we'd want to avoid adding more parameterization there. Seemed to make sense to have `PeerManager` be responsible as a more higher level abstraction for peer-related needs.

Reply:
Related, we'll need to handle `sciddir_or_pubkey` somewhere. That is, going from `Destination` containing a `BlindedPath` with and scid to `OnionMessagePath` having a `Destination` containing a `BlindedPath` with a pubkey.

I suppose doing this in a `MessageRouter` is fine. But when wanting to _construct_ such a blinded path, we'll need the `NetworkGraph`, too. This responsibility then may want to also live in `MessageRouter`, but the current draft is to have it in `Router` partly because `ChannelManager` already has one. So this may need to be rethought.